surface/src/ball_pivoting.cpp

```diff
@@ -0,0 +1,149 @@
+      if (!active_edges_.empty ())
+      {
+        re = boost::make_shared<Edge> (active_edges_.begin ()->second);
+        current_edge_ = re;
+        active_edges_.erase (active_edges_.begin ());
+      }
+      return re;
+    }
+    
+    void
+    BallPivotingFront::addTriangle (const pcl::Vertices &seed, 
+                                    const Eigen::Vector3f &center,
+                                    const bool is_back_ball)
+    {
+      for (size_t idv = 0; idv < 3; ++idv)
+      {
+        addEdge (Edge (seed.vertices.at (idv), seed.vertices.at ((idv + 2) % 3), 
+                       seed.vertices.at ((idv + 1) % 3), center, is_back_ball));
+      }
+    }
+    
+    void
+    BallPivotingFront::addPoint (const Edge &last_edge, const uint32_t id_vertice_extended, 
+                                 const Eigen::Vector3f &center, const bool is_back_ball)
+    {
+      addEdge (Edge (last_edge.id_point_start_, id_vertice_extended,
+                     last_edge.id_point_end_, center, is_back_ball));
+    
+      addEdge (Edge (id_vertice_extended, last_edge.id_point_end_,
+                     last_edge.id_point_start_, center, is_back_ball));
+    }
+    
+    void
+    BallPivotingFront::addEdge (const Edge &edge)
+    {
+      if (!isEdgeOnFront (edge) && !isEdgeFinished (edge))
+      {
+        active_edges_.insert (std::pair<Signature, const Edge> (
+              edge.getSignature (), edge));
+      }
+    }
+    
+    bool
+    BallPivotingFront::isEdgeOnFront (const Edge &edge) const
+    {
+      return active_edges_.find (edge.getSignature ()) != active_edges_.end () ||
+             active_edges_.find (edge.getSignatureReverse ()) != active_edges_.end ();
+    }
+    
+    void
+    BallPivotingFront::removeEdge (const uint32_t id0, const uint32_t id1)
+    {
+      if (!active_edges_.empty ())
+      {
+        typename std::map<Signature, const Edge>::iterator loc = 
+          active_edges_.find (Signature (id0, id1));
+        if (loc != active_edges_.end ())
+        {
+          active_edges_.erase (loc);
+        }
+        else // comment to delete one-directionally
+        {
+          loc = active_edges_.find (Signature (id1, id0));
+          if (loc != active_edges_.end ())
+          {
+            active_edges_.erase (loc);
+          }
+        }
+      }
+    }
+    
+    void
+    BallPivotingFront::clear ()
+    {
+      active_edges_.clear ();
+      finished_signatures_.clear ();
+      current_edge_.reset ();
+    }
+    
+    bool
+    BallPivotingFront::isEdgeFinished (const Edge &edge) const
+    {
+      return finished_signatures_.find (edge.getSignature ()) != finished_signatures_.end () ||
+             finished_signatures_.find (edge.getSignatureReverse ()) != finished_signatures_.end ();
+    }
+  } // namespace pcl::bpa
+} // namespace pcl
+
+// Instantiations of specific point types
+PCL_INSTANTIATE(BallPivoting, (pcl::PointNormal)(pcl::PointXYZRGBNormal)(pcl::PointXYZINormal))
```

Comment:
This should be conditionally compiled depending on the `PCL_NO_PRECOMPILE` macro. Please take [this](https://github.com/PointCloudLibrary/pcl/blob/master/features/src/spin_image.cpp) as an example.